include/envoy/buffer/buffer.h

```diff
@@ -393,10 +393,42 @@ class Instance {
   template <typename T, size_t Size = sizeof(T)> void writeBEInt(T value) {
     writeInt<ByteOrder::BigEndian, T, Size>(value);
   }
+
+  /**
+   * Set the buffer's high watermark. The buffer's low watermark is implicitly set to half the high
+   * watermark. Setting the high watermark to 0 disables watermark functionality.
+   * @param watermark supplies the buffer high watermark size threshold, in bytes.
+   */
+  virtual void setWatermarks(uint32_t watermark) PURE;
+  /**
+   * Returns the configured high watermark.
+   */
+  virtual uint32_t highWatermark() const PURE;
+  /**
+   * Determine if the buffer watermark trigger condition is currently set. The watermark trigger is
+   * set when the buffer size exceeds the configured high watermark and is cleared once the buffer
+   * size drops to the low watermark.
+   * @return true if the buffer size once exceeded the high watermark and hasn't since dropped to
+   * the low watermark.
+   */
+  virtual bool highWatermarkTriggered() const PURE;
 };
 
 using InstancePtr = std::unique_ptr<Instance>;
 
+// Informational enum that hints at the buffer's intended use.
+enum class BufferType {
```

Comment:
Added a comment.  If you're curious, here is the first followup change that depends on this.  The end goal is to use this in e2e tests for the fix to https://github.com/envoyproxy/envoy/issues/11370

https://github.com/envoyproxy/envoy/compare/master...antoniovicente:e2e_h2_high_buffering__pre_shared_ptr_factory

Reply:
ok, so the intended use is to avoid a bunch of nasty test changes and potentially nasty test maintenance issues by switching more buffers to mocks?

I'd like to avoid this if we can.
I wonder if we can reduce the use of the mock buffer factory, and then back this part out.  It looks like it's only really "used" in the tcp proxy integration test so I wonder if we could use a normal buffer factory in the base integration test, fix up the tcp proxy test with some comments (it's not likely to change as frequently as the others) and then drop buffertype from this PR.  WDYT?

Reply:
I'm really not a fan of mock buffers, so no - I am not planning to increase mock buffer use.  I removed the BufferType argument and made the necessary changes to fix broken tests, see https://github.com/envoyproxy/envoy/pull/14256/commits/5b5a99af2d1b0da5e70056949baa8a86ce4a75e0

The change to the default behavior of the mock watermark factory created by the mock dispatcher fixes a bunch of small tests that use MockConnection